src/session_state.rs

```diff
@@ -74,16 +75,34 @@ impl<C: Clock> SessionState<C> {
         }
     }
 
+    pub fn handle_connection_reset(&mut self) {
+        self.last_transmission = None;
+        self.last_reception = None;
+        self.ping_timeout = None;
+    }
+
     pub fn register_transmission(&mut self, now: Instant<C>) {
         self.last_transmission = Some(now);
     }
 
+    pub fn register_reception(&mut self, now: Instant<C>) {
+        self.last_reception = Some(now);
+    }
+
     pub fn ping_is_due(&self, now: &Instant<C>) -> bool {
-        // Send a ping if we haven't sent a transmission in the last 50% of the keepalive internal.
-        self.keep_alive_interval
-            .zip(self.last_transmission)
-            .map_or(false, |(interval, last)| {
-                *now > last + 500.milliseconds() * interval.0
-            })
+        // Send a ping if we haven't sent or received data within 50% of the keep-alive interval.
+        if let Some(interval) = self.keep_alive_interval {
+            let tx_timeout = self
+                .last_transmission
+                .map_or(false, |last| *now > last + 500.milliseconds() * interval.0);
```

Comment:
Took me a while to understand `500.milliseconds() * interval.0`. What's the units of the factors and the units of the product?
`500.milliseconds() * (interval/1.seconds())` would be meaningful to me (if that is possible).
Same in L101.
And who guarantees that the comparison is done in milliseconds. Are `now` and `last` always in milliseconds?

Reply:
This is ultimately an inconvenience imposed by the embedded-time types. Division of `Duration`s use the left-hand type, so dividing 1.seconds() / 2 = 0 seconds, which is unfortunate.

I'm going to just refactor so that we have timeouts 500ms before the deadline - the spec indicates that servers should disconnect clients at 1.5x the timeout interval.